bson.cpp

```diff
@@ -424,6 +428,91 @@ void VariantToBsonConverter::_convertSerializable(bson_t *bson, const char *key,
 }
 /* }}} */
 
+/* {{{ MongoDriver\Driver\ReadConcern */
+void VariantToBsonConverter::_convertReadConcern(bson_t *bson, const char *key, Object v)
+{
+	MongoDBDriverReadConcernData* data = Native::data<MongoDBDriverReadConcernData>(v);
+	const char *rp_string = mongoc_read_concern_get_level(data->m_read_concern);
+	bson_t child;
+
+	bson_append_document_begin(bson, key, -1, &child);
+
+	bson_append_utf8(&child, "level", -1, rp_string, -1);
+
+	bson_append_document_end(bson, &child);
+}
+/* }}} */
+
+/* {{{ MongoDriver\Driver\ReadPreference */
+/* Stolen from libmongoc/src/mongoc/mongoc-read-prefs.c */
+static const char * _get_read_mode_string (mongoc_read_mode_t mode)
+{
+	switch (mode) {
+		case MONGOC_READ_PRIMARY:
+			return "primary";
+		case MONGOC_READ_PRIMARY_PREFERRED:
+			return "primaryPreferred";
+		case MONGOC_READ_SECONDARY:
+			return "secondary";
+		case MONGOC_READ_SECONDARY_PREFERRED:
+			return "secondaryPreferred";
+		case MONGOC_READ_NEAREST:
+			return "nearest";
+		default:
+			return "";
+	}
+}
+
+void VariantToBsonConverter::_convertReadPreference(bson_t *bson, const char *key, Object v)
+{
+	MongoDBDriverReadPreferenceData* data = Native::data<MongoDBDriverReadPreferenceData>(v);
+	const bson_t *tags = mongoc_read_prefs_get_tags(data->m_read_preference);
+	const char *rp_string = _get_read_mode_string(mongoc_read_prefs_get_mode(data->m_read_preference));
+	bson_t child;
+
+	bson_append_document_begin(bson, key, -1, &child);
+
+	bson_append_utf8(&child, "mode", -1, rp_string, -1);
+	if (!bson_empty(tags)) {
+		bson_append_array(&child, "tags", -1, bson_new_from_data(bson_get_data(tags), tags->len));
+	}
```

Comment:
This will need a "maxStalenessMS" field added ( once [HHVM-239](https://jira.mongodb.org/browse/HHVM-239) is implemented. It can be omitted if the value is zero. I'm noting the field name from the spec tests and assuming the rules for omission, as it doesn't appear to be explicitly stated in https://github.com/mongodb/specifications/blob/master/source/max-staleness/.